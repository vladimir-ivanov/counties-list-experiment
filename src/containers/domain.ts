export enum SortBy {
  Name = 'name',
  Population = 'population'
}

export enum SortDirection {
  Asc = 'asc',
  Desc = 'desc'
}

export interface SortField {
  field: SortBy | undefined,
  direction: SortDirection | undefined
}

