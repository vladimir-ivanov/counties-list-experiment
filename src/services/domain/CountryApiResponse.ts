export interface CountryApiResponse {
  name: string;
  alpha3Code: string;
  capital: string;
  population: number;
  currencies: Currency[];
  languages: Language[];
}

export interface Currency {
  code: string;
  name: string;
  symbol: string;
}

export interface Language {
  iso639_1: string;
  iso639_2: string;
  name: string;
  nativeName: string;
}
