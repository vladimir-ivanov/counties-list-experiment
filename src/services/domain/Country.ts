import { CountryApiResponse } from './CountryApiResponse';

type PickProps = 'name' | 'alpha3Code' | 'capital' | 'population';

export interface Country extends Pick<CountryApiResponse, PickProps> {
  currencies: string[];
  languages: string[];
}
