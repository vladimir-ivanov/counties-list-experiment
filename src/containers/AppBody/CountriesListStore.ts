import { Country } from '../../services/domain/Country';
import { SortDirection, SortField } from '../domain';

export class CountriesListStore {
  private readonly countriesList: Country[];
  private sortField: SortField = {} as SortField;
  private term: string = '';

  constructor(countries: Country[]) {
    this.countriesList = countries;
  }

  setSortField = (sortField: SortField) => {
    this.sortField = sortField;

    return this;
  };

  setTerm = (term: string) => {
    this.term = term;

    return this;
  };

  getUpdatedCountries = () => {
    let countries =  this.countriesList;

    if (this.sortField) {
      const {direction, field} = this.sortField;

      countries = [...countries].sort((a, b) => {
        // todo sort these boys out
        // @ts-ignore
        const fieldA = a[field] as any;
        // @ts-ignore
        const fieldB = b[field] as any;
        const directionIsAsc = direction === SortDirection.Asc;

        if (fieldA > fieldB) {
          return directionIsAsc ? 1 : -1;
        }

        if (fieldA < fieldB) {
          return directionIsAsc ? -1 : 1;
        }

        return 0;
      });
    }

    if (this.term) {
      countries = countries.filter(({ name }) => name.toLowerCase().indexOf(this.term) !== -1);
    }

    return countries;
  };
}