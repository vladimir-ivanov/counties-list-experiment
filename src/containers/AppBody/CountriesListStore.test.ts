import { CountriesListStore } from './CountriesListStore';
import { SortBy, SortDirection } from '../domain';

describe('CountriesListStore', () => {
  const countryListStore = new CountriesListStore([
    {
      'currencies': [],
      'languages': [],
      'name': 'Afghanistan',
      'alpha3Code': 'AFG',
      'capital': 'Kabul',
      'population': 1
    },
    {
      'currencies': [],
      'languages': [],
      'name': 'UK',
      'alpha3Code': 'GBR',
      'capital': 'London',
      'population': 2
    },
    {
      'currencies': [],
      'languages': [],
      'name': 'USA',
      'alpha3Code': 'USA',
      'capital': 'Washington DC',
      'population': 3
    }
  ]);


  describe('execute()', () => {
    describe('withNoTermOrSort', () => {
      it('should return the input list as is', () => {
        expect(countryListStore.execute()).toMatchSnapshot();
      });
    });

    describe('withTermNoSort', () => {
      it('should return the input list matches only', () => {
        expect(countryListStore.setTerm('u').execute()).toMatchSnapshot();
      });
    });

    // ideally would need to test all scenarios if no time constraints to finish this test
    describe('withSortNoTerm', () => {
      it('should return the input list sorted by population', () => {
        expect(
          countryListStore
            .setSortField({
              field: SortBy.Population,
              direction: SortDirection.Desc
            })
            .execute()
        ).toMatchSnapshot();
      });
    });

    describe('withSortAndTerm', () => {
      it('should return the input list sorted by population', () => {
        expect(
          countryListStore
            .setSortField({
              field: SortBy.Population,
              direction: SortDirection.Desc
            })
            .setTerm('u')
            .execute()
        ).toMatchSnapshot();
      });
    });
  });
});