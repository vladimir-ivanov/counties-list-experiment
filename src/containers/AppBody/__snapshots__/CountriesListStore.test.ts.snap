// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountriesListStore execute() withNoTermOrSort should return the input list as is 1`] = `
Array [
  Object {
    "alpha3Code": "AFG",
    "capital": "Kabul",
    "currencies": Array [],
    "languages": Array [],
    "name": "Afghanistan",
    "population": 1,
  },
  Object {
    "alpha3Code": "GBR",
    "capital": "London",
    "currencies": Array [],
    "languages": Array [],
    "name": "UK",
    "population": 2,
  },
  Object {
    "alpha3Code": "USA",
    "capital": "Washington DC",
    "currencies": Array [],
    "languages": Array [],
    "name": "USA",
    "population": 3,
  },
]
`;

exports[`CountriesListStore execute() withSortAndTerm should return the input list sorted by population 1`] = `
Array [
  Object {
    "alpha3Code": "USA",
    "capital": "Washington DC",
    "currencies": Array [],
    "languages": Array [],
    "name": "USA",
    "population": 3,
  },
  Object {
    "alpha3Code": "GBR",
    "capital": "London",
    "currencies": Array [],
    "languages": Array [],
    "name": "UK",
    "population": 2,
  },
]
`;

exports[`CountriesListStore execute() withSortNoTerm should return the input list sorted by population 1`] = `
Array [
  Object {
    "alpha3Code": "USA",
    "capital": "Washington DC",
    "currencies": Array [],
    "languages": Array [],
    "name": "USA",
    "population": 3,
  },
  Object {
    "alpha3Code": "GBR",
    "capital": "London",
    "currencies": Array [],
    "languages": Array [],
    "name": "UK",
    "population": 2,
  },
  Object {
    "alpha3Code": "AFG",
    "capital": "Kabul",
    "currencies": Array [],
    "languages": Array [],
    "name": "Afghanistan",
    "population": 1,
  },
]
`;

exports[`CountriesListStore execute() withTermNoSort should return the input list matches only 1`] = `
Array [
  Object {
    "alpha3Code": "GBR",
    "capital": "London",
    "currencies": Array [],
    "languages": Array [],
    "name": "UK",
    "population": 2,
  },
  Object {
    "alpha3Code": "USA",
    "capital": "Washington DC",
    "currencies": Array [],
    "languages": Array [],
    "name": "USA",
    "population": 3,
  },
]
`;
